src/core/tsi/ssl_transport_security.cc

```diff
@@ -490,21 +490,34 @@ static tsi_result peer_from_x509(X509* cert, int include_certificate_type,
   return result;
 }
 
-/* Logs the SSL error stack. */
-static void log_ssl_error_stack(void) {
-  unsigned long err;
-  while ((err = ERR_get_error()) != 0) {
-    char details[256];
-    ERR_error_string_n(static_cast<uint32_t>(err), details, sizeof(details));
-    gpr_log(GPR_ERROR, "%s", details);
+/* Clears all SSL errors from the stack, and returns them as a string. */
+static std::string ssl_error_stack_to_string(void) {
+  BIO* bio = BIO_new(BIO_s_mem());
+  ERR_print_errors(bio);
+
+  const uint8_t* buf;
+  size_t buf_len;
+  std::string result;
+  if (!BIO_mem_contents(bio, &buf, &buf_len) || buf_len == 0) {
+    result = "error could not be retrieved from the SSL stack";
+  } else {
+    result = std::string(reinterpret_cast<const char*>(buf), buf_len);
   }
+  BIO_free(bio);
+  return result;
 }
 
 /* Performs an SSL_read and handle errors. */
 static tsi_result do_ssl_read(SSL* ssl, unsigned char* unprotected_bytes,
                               size_t* unprotected_bytes_size) {
-  int read_from_ssl;
   GPR_ASSERT(*unprotected_bytes_size <= INT_MAX);
+  // Gracefully handle the case where some other part of the application left
+  // an error on the stack.
+  if (ERR_peek_error() != 0) {
+    gpr_log(GPR_ERROR, "do_ssl_read invoked with a dirty SSL error stack: %s",
+            ssl_error_stack_to_string().c_str());
+  }
+  int read_from_ssl;
   read_from_ssl = SSL_read(ssl, unprotected_bytes,
```

Comment:
There are many more places in this file than `SSL_read` which are paired with `SSL_get_error`. If we want to do this in gRPC, we should do it in front of the other ones too. Hence helper function above.